### PR TITLE
Update QUICHE from 9386be8fb to d0d4d290e

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
Update QUICHE from 9386be8fb to d0d4d290e
https://github.com/google/quiche/compare/9386be8fb..d0d4d290e
    
$ git log 9386be8fb..d0d4d290e --date=short --no-merges --format="%ad %al %s"
    
2022-03-15 quiche-dev Swap QUIC_LOG to QUIC_DLOG.
2022-03-15 diannahu Handle outbound SETTINGS_MAX_FRAME_SIZE in oghttp2.
2022-03-15 bnc Move QuicEpollClock to quic/core.
2022-03-14 wub Close QUIC connection if it fails to serialize a coalesced packet.
2022-03-14 dschinazi Deprecate quic_version_information flag
2022-03-14 diannahu Remove Http2DecoderAdapter::Reset() and ResetInternal().
2022-03-14 martinduke Server ID object for QUIC-LB. Wraps an array of uint8_t with validation to make sure the length is correct. This will be used by other classes.
2022-03-14 bnc Internal change
2022-03-10 quiche-dev Add Http2DecoderAdapter::SetMaxFrameSize(), currently unused.
2022-03-10 bnc Move mutex and testvalue from quic/platform/impl to quiche/common/platform/impl.
    
Signed-off-by: Ryan Hamilton <rch@google.com>

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
/lgtm deps

**Comment by daixiang0** [comment]
I have seen some PRs related to QUICHE update, do we need always to update?

/cc @mattklein123

**Comment by RyanTheOptimist** [comment]
> I have seen some PRs related to QUICHE update, do we need always to update?
> 
> /cc @mattklein123

Yes, we need to update in order to take advantage of the ongoing QUICHE development which powers our QUIC and HTTP stacks.

File: source/common/quic/platform/quiche_mutex_impl.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#pragma once
```
Comment (bencebeky) [inline source/common/quic/platform/quiche_mutex_impl.h:1]: This file looks identical to https://github.com/google/quiche/blob/main/common/platform/default/quiche_platform_impl/quiche_mutex_impl.h with method definitions in https://github.com/google/quiche/blob/main/common/platform/default/quiche_platform_impl/quiche_mutex_impl.cc inlined.  Could it be removed just like quic_testvalue_impl.h was removed?

File: bazel/external/quiche.BUILD [hunk 2]
```diff
@@ -4207,13 +4262,11 @@ envoy_cc_library(
 envoy_cc_library(
     name = "quiche_common_platform_default_quiche_platform_impl_lib",
     srcs = [
-        "quiche/common/platform/default/quiche_platform_impl/quic_mutex_impl.cc",
         "quiche/common/platform/default/quiche_platform_impl/quiche_time_utils_impl.cc",
     ],
     hdrs = [
-        "quiche/common/platform/default/quiche_platform_impl/quic_mutex_impl.h",
```
Comment (bencebeky) [inline bazel/external/quiche.BUILD:4214]: I don't understand why there was a build target with platform defaults for mutex and testvalue here, while :quic_platform_base_impl_lib had Envoy's quic_testvalue_impl.h and :quic_platform_impl_lib had Envoy's quic_mutex_impl.h.  I would assume that any QUICHE consumer would only need to build one or the other (default in QUICHE or local impl), not both.  It feels like a pre-existing inconsistency (maybe leftover from when default implementation was added) and I support fixing it.

File: bazel/external/quiche.BUILD [hunk 3]
```diff
@@ -1248,6 +1272,7 @@ envoy_cc_test_library(
     repository = "@envoy",
     tags = ["nofips"],
     deps = [
+        ":quiche_common_platform_mutex",
```
Comment (bencebeky) [inline bazel/external/quiche.BUILD:1275]: Why does :quic_platform_test depend on :quiche_common_platform_mutex?  (It doesn't in Google3.)

File: source/common/quic/platform/BUILD [hunk 4]
```diff
@@ -122,28 +121,38 @@ envoy_cc_library(
...
-        "//source/common/common:assert_lib",
-        "//source/common/filesystem:directory_lib",
-        "//source/common/filesystem:filesystem_lib",
         "//source/common/http:utility_lib",
+        "@com_github_google_quiche//:quic_platform_export",
+    ],
+)
+
+envoy_cc_library(
+    name = "quic_platform_mutex_impl_lib",
```
Comment (bencebeky) [inline source/common/quic/platform/BUILD:144]: Consider renaming this target to :quiche_platform_mutex_impl_lib.

**Review by bencebeky — APPROVED** [review]

File: bazel/external/quiche.BUILD (hunk 3 above)
Comment (RyanTheOptimist) [reply bazel/external/quiche.BUILD]: Removed. I thought I needed it to, but clearly not.

File: source/common/quic/platform/BUILD (hunk 4 above)
Comment (RyanTheOptimist) [reply source/common/quic/platform/BUILD]: I've removed this entire target now that we're using the common impl (as per your great observation!).

File: source/common/quic/platform/quiche_mutex_impl.h (hunk 1 above)
Comment (RyanTheOptimist) [reply source/common/quic/platform/quiche_mutex_impl.h]: AH! Yes, good point. Done.

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks, Bence!

**Review by alyssawilk — APPROVED** [review]
Good luck with CI :-/

**Comment by ggreenway** [comment]
> @alyssawilk the tsan failure is in //test/integration:h1_capture_persistent_fuzz_test which seems utterly unrelated to this. I'm happy to re-test as I suspect this is a flake. But GH is crazy slow today. What do you think of landing with this failure?
> 
> (In any case, please review the PR :>)

There was another instance of this earlier today. Definitely unrelated to this PR. Filed #20382 to track.

**Pull request merged**